torch/csrc/autograd/record_function.h

```diff
@@ -0,0 +1,109 @@
+#pragma once
+
+#include <ATen/core/ivalue.h>
+#include <c10/util/SmallVector.h>
+
+namespace torch { namespace autograd {
+
+struct Function;
+
+namespace profiler {
+
+using GetPackedInputsCallback = std::function<std::vector<c10::IValue>()>;
+
+struct FunctionCallContext {
```

Comment:
I am not sure we are gaining much from separating this from RecordFunction. It causes more allocations and refcounting to support recording through asynchronous calls, but it is not clear to me we really should be recording async trace information through forks anyway. If we have a different hook for profiling the interpreter anyway, then it would make more sense for that facility to be able to collect accurate interpreter stack traces.

Reply:
One specific use case I had in mind - associating an op with a specific script. Interpreter knows what code it runs, while RecordFunction does the (sampled) logging and doesn't have the interpreter context.
If I understand your suggestion correctly, technically we could set the "starting" thread local logging info for the new thread (copied from the parent thread) and use raw pointers to link nested RecordFunctions.

Reply:
That would remove necessity for refcounting and we could use c10::optional<FunctionCallContext> object as a member of RecordFunction. I guess from the design point of view one argument to keep FunctionCallContext is to control information passed to the callbacks by aggregating it in FunctionCallContext.